### PR TITLE
expression: handle zero date in `date_add()`

### LINKED ISSUE #15234: zeroDate behavior for date_add and timestampdiff is not consistent in tidb
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

mysql> select timestampdiff(day,'0000-00-00 00:00:00','2020-01-01 00:00:00');
+----------------------------------------------------------------+
| timestampdiff(day,'0000-00-00 00:00:00','2020-01-01 00:00:00') |
+----------------------------------------------------------------+
|                                                           NULL |
+----------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select date_add('0000-00-00 00:00:00', interval 2020 year);
+-----------------------------------------------------+
| date_add('0000-00-00 00:00:00', interval 2020 year) |
+-----------------------------------------------------+
| 2019-11-30 00:00:00                                 |
+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

2. What did you expect to see?
timestampdiff and date_add should both return NULL or meaningful result.
BTW in mysql both timestampdiff and date_add return NULL

3. What did you see instead?

mysql> select timestampdiff(day,'0000-00-00 00:00:00','2020-01-01 00:00:00');
+----------------------------------------------------------------+
| timestampdiff(day,'0000-00-00 00:00:00','2020-01-01 00:00:00') |
+----------------------------------------------------------------+
|                                                           NULL |
+----------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select date_add('0000-00-00 00:00:00', interval 2020 year);
+-----------------------------------------------------+
| date_add('0000-00-00 00:00:00', interval 2020 year) |
+-----------------------------------------------------+
| 2019-11-30 00:00:00                                 |
+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

tidb master

### PR DESCRIPTION
**Pull request by zhangysh1995**
### What problem does this PR solve?

Issue Number:  close #15234 

Problem Summary: 
*  Invalid zero date could be processed

````
mysql> select date_add('0000-00-00 00:00:00', interval 2020 year);
+-----------------------------------------------------+
| date_add('0000-00-00 00:00:00', interval 2020 year) |
+-----------------------------------------------------+
| 2019-11-30 00:00:00                                 |
+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------+
| Level   | Code | Message                                 |
+---------+------+-----------------------------------------+
| Warning | 1292 | Incorrect time value: '{0 0 0 0 0 0 0}' |
+---------+------+-----------------------------------------+
1 row in set (0.00 sec)

````

### What is changed and how it works?

What's Changed:

*  Return `NULL` when the first date is invalid

How it Works:

*  Add a invalid check to how related functions

```` golang
        ...
	if invalidDate := date.InvalidZero(); invalidDate {
		err = handleInvalidTimeError(b.ctx, types.ErrWrongValue.GenWithStackByArgs(types.TimeStr, date.String()))
		return types.ZeroTime, true, err
	}
        ...
````

* Now it becomes:
````
mysql> select date_add('0000-00-00 00:00:00', interval 2020 year);
+-----------------------------------------------------+
| date_add('0000-00-00 00:00:00', interval 2020 year) |
+-----------------------------------------------------+
| NULL                                                |
+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------------------+
| Level   | Code | Message                                            |
+---------+------+----------------------------------------------------+
| Warning | 1292 | Incorrect time value: '0000-00-00 00:00:00.000000' |
+---------+------+----------------------------------------------------+
1 row in set (0.00 sec
````

* Think a refactor of this patch is necessary, because I use paste now.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test

### Release note 
* DATE_ADD is compatible with MySQL with handling zero date

### CODE REVIEW

**Comment by zhangysh1995** [comment]
@ichn-hu Could you please have a look at the code?

**Review by rebelice — COMMENTED** [review]
Could you please add some tests for this PR?

**Comment by ichn-hu** [comment]
> @ichn-hu Could you please have a look at the code?

Sure, sorry for the delay, I don't see any thing wrong, but would you please as @rebelice said add some test which reflects  both `timestampdiff` and `date_add` returned `NULL` on the given test case?

**Comment by zhangysh1995** [comment]
I found that when the second argument is invalid, TiDB miss an warning:

````
-- mysql 5.7
mysql> select adddate("2018-08-16 20:21:01", "-00:00:00.000000");
+----------------------------------------------------+
| adddate("2018-08-16 20:21:01", "-00:00:00.000000") |
+----------------------------------------------------+
| 2018-08-16 20:21:01                                |
+----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------------------------------+
| Level   | Code | Message                                               |
+---------+------+-------------------------------------------------------+
| Warning | 1292 | Truncated incorrect INTEGER value: '-00:00:00.000000' |
+---------+------+-------------------------------------------------------+
1 row in set (0.00 sec)

-- TiDB
mysql> select adddate("2018-08-16 20:21:01", "-00:00:00.000000");
+----------------------------------------------------+
| adddate("2018-08-16 20:21:01", "-00:00:00.000000") |
+----------------------------------------------------+
| 2018-08-16 20:21:01                                |
+----------------------------------------------------+
1 row in set (0.00 sec)
````
Because this is not related to the current issue, I will not fix it here.

**Comment by zhangysh1995** [comment]
@XuHuaiyu What would be a possible reward for this issue?

**Comment by zhangysh1995** [comment]
@rebelice @ichn-hu Added some test~ Basically, I didn't see a test for `date_add` so I created a new one.

File: expression/builtin_time.go [hunk 1]
```diff
@@ -3416,6 +3416,11 @@ func (b *builtinAddDateStringStringSig) evalTime(row chunk.Row) (types.Time, boo
 		return types.ZeroTime, true, err
```
Comment (XuHuaiyu) [inline expression/builtin_time.go:3416]: VecEvalTime should also be modified, you can check the following example in MySQL and TiDB
``` sql
set @@sql_mode='';
create table t(a datetime);
insert into t values('0000-00-00 00:00:00');
select date_add(a, interval 2020 year) from t;
```

But it'll be ok to fix this in another PR.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -3449,6 +3454,11 @@ func (b *builtinAddDateStringIntSig) evalTime(row chunk.Row) (types.Time, bool,
...
 	}
 
+	if invalidDate := date.InvalidZero(); invalidDate {
```
[the remark this answers is not in the document]
Comment (zhangysh1995) [reply expression/builtin_time.go:3457]: I think it will help to remove the redundant check in the caller function, but it depends on the context to determine whether zeroDate is valid. For example, `adddate('0000-00-00', interval 1 day)`  is invalid becuase the first zero, while `adddate('2000-10-11', 0)` is valid with the second argument as a zero. It should be better leave to the caller function, but my current fix is not elegant.

File: expression/builtin_time.go (hunk 1 above)
Comment (zhangysh1995) [reply expression/builtin_time.go:3416]: @XuHuaiyu @SunRunAway I think this vectorized function should also be fixed, but it may be too many changes for a single PR. To have a cleaner handling of this issue, a refactor of the code may be necessary. For example, I put the valid check in all the `add_date` functions, but I saw that there is _little differences_ among all the implemented functions. If you think it is better to fix it in the same PR, I will do it. But I will prefer to open a new one for it.

**Comment by XuHuaiyu** [comment]
I think that handle the vec-cases in another PR is acceptable.

**Review by ichn-hu — APPROVED** [review]
LGTM

**Comment by wshwsh12** [comment]
In master branch 49791bc3f, 
```
[tidb]> select date_add('0000-00-00 00:00:00', interval 2020 year);
+-----------------------------------------------------+
| date_add('0000-00-00 00:00:00', interval 2020 year) |
+-----------------------------------------------------+
| NULL                                                |
+-----------------------------------------------------+
1 row in set, 1 warning (0.000 sec)
```
So the problem has been fixed?

**Review by wshwsh12 — COMMENTED** [review]
Why not add error handle in getDateFromXXXXXX ang vecGetDateFromXXXXXX?

**Comment by zhangysh1995** [comment]
It has been a while since I submitted this PR, I will take a look recently.

**Comment by zz-jason** [comment]
@zhangysh1995 any update?

**Comment by zhangysh1995** [comment]
> @zhangysh1995 any update?

I think a clarification on the scope of this PR is needed. First I want to fix for the non-vectorization case, then the vectorized case is required. Later because the master is fixed, and the latest review requires a refactor of my fix. I think I need a clear goal now for this PR. If a refactor and the additional feature should be fixed, I would consider close this one and open another. @wshwsh12 @XuHuaiyu What are your comments?

**Comment by tisonkun** [comment]
@zhangysh1995 from @wshwsh12 's comment I think this PR doesn't fix the issue. There are still two unanswered questions:

* https://github.com/pingcap/tidb/pull/20015#issuecomment-749896837
* https://github.com/pingcap/tidb/pull/20015#pullrequestreview-557486283

You can update the PR to fix the specific issue reported in https://github.com/pingcap/tidb/issues/15234, regardless any further requirement.

**Comment by zhangysh1995** [comment]
需求变动比较大，现在没精力改这个 PR 了，希望重新 assign 。
